### PR TITLE
TRITON-2202 Document RFD 176 in "triton" repo

### PR DESCRIPTION
**Pull request by danmcd**

### CODE REVIEW

File: docs/developer-guide/ipxe-installer.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+reachability be enabled DURING the installation process.  The installer
+script will post a reminder of this.  Most of the time, the iPXE installer
+needed network reachability to start in the first place, so configuring the
+head node to reach the iPXE downloads should not be difficult.  The network
+is required to be able to download images that are normally on the ISO DVD
+image for the ISO installer, but cannot fit into an iPXE ramdisk root.
+
+## Configuring an iPXE Server for a Triton Installation
+
+A Triton iPXE server needs to be populated so iPXE can pull via HTTP what it
```
Comment (jasonbking) [inline docs/developer-guide/ipxe-installer.md:32]: This sentence is hard to parse. Rewrite?

File: docs/developer-guide/iso-installer.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+-->
+
+<!--
+    Copyright 2021, Joyent, Inc.
+-->
+
+# Triton ISO Installation
+
+If one wishes to boot off a ZFS pool, one can use the Triton ISO installer to
+boot once off of DVD-ROM, and on subsequent boots boot off of a bootable
```
Comment (jasonbking) [inline docs/developer-guide/iso-installer.md:14]: Maybe a comma after 'on subsequent boots' ?

File: docs/developer-guide/iso-installer.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+See the [main README's Physical Server
+section](https://github.com/joyent/triton/tree/TRITON-2202#installing-triton-on-a-physical-server)
+for more details
+
+## Disk/Pool Requirements
+
+See [Booting the Head Node from a ZFS
+Pool](https://github.com/joyent/triton/blob/TRITON-2202/docs/developer-guide/zpool.md)
+for details about how to
```
Comment (jasonbking) [inline docs/developer-guide/iso-installer.md:40]: 'how to ...?'

File: docs/developer-guide/iso-installer.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+See [Booting the Head Node from a ZFS
+Pool](https://github.com/joyent/triton/blob/TRITON-2202/docs/developer-guide/zpool.md)
+for details about how to 
+
+## Boot-device Requirements
+
+It is important that the Triton ISO installer is only booted once.  Part of
+the Triton installation process is to reboot a second time, which in the case
+of the ISO installer must be to the boot disks.  The best way to configure
+this is to have hard-drives be booted first, but the DVD drive be the
```
Comment (jasonbking) [inline docs/developer-guide/iso-installer.md:47]: Maybe 'have hard drives first in the boot order followed by the DVD drive second (or later) in the boot order' ?

File: docs/developer-guide/zpool.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+- Two or more same-sized disk with a single SSD for a log device.
+
+NOTE:  After installation a bootable pool can have log devices or log device
+mirrors added.
+
+## Installing a ZFS bootable Head Node
+
+In addition to [regular network attachment and
+configuration](https://docs.joyent.com/private-cloud/install/network-layout),
+if you wish to start booting a Head Node from a ZFS pool you should attach
```
Comment (jasonbking) [inline docs/developer-guide/zpool.md:48]: 'you should attach set of server disks' -- I think you a word :)

File: docs/developer-guide/zpool.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+NOT be the first tried boot device (this is ESPECIALLY true for virtual
+machines).  Also, the server (or virtual machine) should have at least two
+VLANs or actual LANs associated with its ethernet port(s).  If you are
+installing a disk-bootable CoaL, you can either follow the CoaL directions to
+setup external and admin network, OR you may configure your own prior to
+booting the ISO installer.
+
+Make sure the Triton installation ISO is in the optical drive (virtual or
+otherwise), and boot it.  Proceed to follow the installation prompts per
+[Installing Triton Elastic Container
```
Comment (jasonbking) [inline docs/developer-guide/zpool.md:70]: Is that the correct title for the linked document?

File: docs/developer-guide/zpool.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+Infrastructure](https://docs.joyent.com/private-cloud/install).
+
+### For iPXE Installations
+
+As with ISO installations, it is important that the boot order for the server
```
Comment (jasonbking) [inline docs/developer-guide/zpool.md:75]: 'it is important that the disk are first in the boot order' I think sounds a bit better.

File: docs/developer-guide/zpool.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+An iPXE installation requires that iPXE web directory that allows a boot into
+the Triton installer also be reachable by the configured-at-installation time
+"external" network.  The iPXE web directory contains a copy of the larger ISO
+contents, which cannot fit into an iPXE boot-archive/initrd.
+
+
+## Converting from a USB-key (or from one bootable pool to another)
+
+The [piadm(1M)
+command](https://github.com/joyent/smartos-live/blob/master/man/usr/share/man/man1m/piadm.1m.md)
```
Comment (jasonbking) [inline docs/developer-guide/zpool.md:88]: Maybe just have `piadm(1M)` be the link and 'command' just regular text? (nit)

File: docs/developer-guide/zpool.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+can be employed to transfer the current USB-key to a bootable ZFS pool.  It
+can also be used to transfer boot contents from one pool to another.  If a
+pool is not bootable, piadm(1M) will fail.
+
+Select a POOL, and then issuing `piadm bootable -e POOL` will copy the USB
```
Comment (jasonbking) [inline docs/developer-guide/zpool.md:93]: Maybe 'Select a POOL, then issue `piadm bootable -e POOL` to copy the USB key contents... ... pool to be bootable. If the pool is not bootable, the `piadm` command will fail.'

File: docs/operator-guide/server-setup.md [hunk 10]
```diff
@@ -28,10 +38,11 @@ the point where the new empty CN first boots.
...
-When a CN boots (whether by PXE or USB) the first interaction it will have with
-the Triton control plane will be via DHCP. It will do a DHCPDISCOVER in order to
-attempt to learn its own IP address. This DHCPDISCOVER is a broadcast message
-and the booter/dhcpd service (usually dhcpd0 zone on the HN) will see this
+When a CN boots (whether by PXE, USB, or from disk) the first
+interaction it will have with the Triton control plane will be via
+DHCP. It will do a DHCPDISCOVER in order to attempt to learn its own
+IP address. This DHCPDISCOVER is a broadcast message and the
+booter/dhcpd service (usually dhcpd0 zone on the HN) will see this
 message be responsible for handling it.
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:46]: Not your typo but 'see this message be responsible' --- I think 'and be responsible' is probably what was meant (if you want to fix).

File: docs/operator-guide/server-setup.md [hunk 11]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
 pre-existing zpool on the disks. When this happens often the server-setup
 plows forward and the agentsetup fails. If you see a failure, it's always a
 good idea to check the status of the zpool early in the debugging process.
+
+## iPXE booting Compute Nodes
+
+In some situations, such as a Bare-Metal-as-a-Service (BMaaS) provider, they
+may have instances that by default only network boot on one network
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:438]: This sounds a bit awkward.. maybe 'they may have instances where by default a system can only network boot on one network interface. The network interface may not be the interface a Triton installation has configured.....'

File: docs/operator-guide/server-setup.md [hunk 12]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
...
+interface, and it may not be the one a Triton installation has configured as
+the `admin` network.  Often times, these single-network-interfaces default to
+an iPXE boot if it can network boot at all.
+
+A suggested course of action in these situations is to, if possible, ALWAYS
+boot using iPXE, but chainloading into a Triton Compute Node boot.
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:444]: Maybe 'chainload' instead?

File: docs/operator-guide/server-setup.md [hunk 13]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
...
+#!ipxe
+set base-url https://example.com/triton-cn-ipxe
+kernel ${base-url}/ipxe.lkrn
+module ${base-url}/default-no0.ipxe
+boot
+```
+
+And the URL for the BMaaS iPXE should be, in this example,
+`https://example.com/triton-cn-ipxe/triton-cn.ipxe`.  There are two other
+files mentioned in this, and those are what we need as well.
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:466]: Maybe 'and those are needed as well'?

File: docs/operator-guide/server-setup.md [hunk 14]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
...
+
+### An iPXE binary
+
+In a head node, the file `/opt/smartdc/share/usbkey/contents/boot/ipxe.lkrn`
+exists, and is the Triton-special iPXE binary that a Triton Compute Node
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:471]: Either 'exists and is the...' or 'exists, and it is the....'

File: docs/operator-guide/server-setup.md [hunk 15]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
...
+normally boots into either off of a USB key or off of a BIOS PXE chainload.
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:472]: Maybe 'normally boots into from a USB key or from a BIOS PXE chainload' ?

File: docs/operator-guide/server-setup.md [hunk 16]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
...
+For a BMaaS compute node that can only network-boot from the BMaaS iPXE, it
+will have to chain load into this one.
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:474]: 'will have to chainload into the Triton-special iPXE binary'

File: docs/operator-guide/server-setup.md [hunk 17]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
...
+
+### The Triton-specific iPXE instruction file
+
+In this example, we name it `default-no0.ipxe`, and it is designed to bypass
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:478]: 'we name the iPXE instruction file ...'

File: docs/operator-guide/server-setup.md [hunk 18]
```diff
@@ -420,3 +431,71 @@ Another common failure mode is to have the joysetup.sh script fail due to a
...
+at least the primary NIC on the compute node when trying to reach the Triton
+Head Node DHCP server.
+
+Some BMaaS providers set `net2` as the primary, because of possible
+dissatisfaction with the on-board NICs, and the desire to use a well-tested
```
Comment (jasonbking) [inline docs/operator-guide/server-setup.md:483]: No comma after NICs

File: README.md [hunk 19]
```diff
@@ -80,36 +80,59 @@ Community discussion about Triton DataCenter happens in two main places:
...
+The [CoaL Setup document](./docs/developer-guide/coal-setup.md) contains
+detailed instructions for downloading and installing the virtual appliance.
+If you use the ISO installer (see below) on a Virtual Machine, the CoaL setup
+instructions post-booting should be followed as well.
 
-If you already have a CoaL and would like to update the installation, follow the instructions
-for [updating a Triton standup using `sdcadm`](https://github.com/joyent/sdcadm/blob/master/docs/update.md).
+If you already have a CoaL and would like to update the installation, follow
+the instructions for [updating a Triton standup using
+`sdcadm`](https://github.com/joyent/sdcadm/blob/master/docs/update.md).
```
Comment (papertigers) [inline README.md:94]: neat! I didn't know github markdown let you use back ticks in a URL link.

File: README.md [hunk 20]
```diff
@@ -80,36 +80,59 @@ Community discussion about Triton DataCenter happens in two main places:
...
+with a ZFS bootable filesystem substituting for the USB-key.
 
 
 #### Hardware
 
 For Triton development only, the minimum server hardware is:
 
 - 8 GB USB flash drive
-- Intel Processors with VT-x and EPT support (all Xeon since Nehalem)
+- Intel Processors with VT-x and EPT support (all Xeon since Nehalem), or AMD Processors no earlier than EPYC or Zen.
```
Comment (papertigers) [inline README.md:127]: Is that true for AMD? I assume this is for if one want's to test bhyve.  `SVM` came in before Zen but I don't know what the minimum requirement for bhyve actually is.

File: docs/developer-guide/ipxe-installer.md [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+needs to boot the Triton installer and actually perform the installation.  We
+provide a [tar
+archive](https://us-east.manta.joyent.com/Joyent_Dev/public/SmartDataCenter/ipxe-latest.tgz)
+that can be installed on a web server that an iPXE client can reach.
+
+The `triton-installer.ipxe` file will need modifications in its `testdomain`
+and `base-url` variables to match your deployment.  If your iPXE webserver is
```
Comment (papertigers) [inline docs/developer-guide/ipxe-installer.md:39]: potentially not needed anymore based on Brian's comments?

File: docs/developer-guide/iso-installer.md (hunk 2 above)
Comment (papertigers) [inline docs/developer-guide/iso-installer.md:14]: comma between "boots, boot"?

File: docs/developer-guide/iso-installer.md [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+and follow the main README. 
+
+## Minimum Requirements
+
+The ISO installer requires more disk space than a USB-key Triton deployment,
+but is otherwise the same.
+
+- Intel Processors with VT-x and EPT support (all Xeon since Nehalem), or AMD Processors no earlier than EPYC or Zen. 
+- 16 GB RAM
+- 60 GB available storage
```
Comment (papertigers) [inline docs/developer-guide/iso-installer.md:30]: Same question as above. Also why the difference between recommended disk sizes?

File: docs/developer-guide/zpool.md [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+
+Not all pools can be bootable.  We suggest a simple disk set for a
+ZFS-bootable Head Node.  Such sets include:
+
+- Single disk.
+- Two (mirrored) or more (raidz) same-sized disk.
+- Two or more same-sized disk with a single SSD for a log device.
+
+NOTE:  After installation a bootable pool can have log devices or log device
+mirrors added.
```
Comment (papertigers) [inline docs/developer-guide/zpool.md:42]: Any reason to call out SLOG devices and not L2ARC? I assume it's also safe?

File: docs/developer-guide/zpool.md [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+
+Make sure the Triton installation ISO is in the optical drive (virtual or
+otherwise), and boot it.  Proceed to follow the installation prompts per
+[Installing Triton Elastic Container
+Infrastructure](https://docs.joyent.com/private-cloud/install).
+
+### For iPXE Installations
+
+As with ISO installations, it is important that the boot order for the server
+be the disks first, but the disks should not be bootable for the initial iPXE
```
Comment (papertigers) [inline docs/developer-guide/zpool.md:76]: Perhaps mention a way to clear the zpool of being bootable if a user wishes to reinstall an environment or VM. (You can also say that's out of scope for this document.  I am cool with that answer).

File: README.md (hunk 20 above)
Comment (danmcd) [reply README.md:127]: I figured I'd rather be conservative here, and yes BHYVE crossed my mind when I settled on this.

File: docs/developer-guide/ipxe-installer.md (hunk 1 above)
Comment (danmcd) [reply docs/developer-guide/ipxe-installer.md:32]: Will do.

File: docs/developer-guide/ipxe-installer.md (hunk 21 above)
Comment (danmcd) [reply docs/developer-guide/ipxe-installer.md:39]: Yeah, save the ISO image URL... will let @bahamat be the Final Word on this.

File: docs/developer-guide/iso-installer.md (hunk 2 above)
Comment (danmcd) [reply docs/developer-guide/iso-installer.md:14]: Ooof, good catch, will fix.

Comment (danmcd) [reply docs/developer-guide/iso-installer.md:14]: @jasonbking got it first.  :)

File: docs/developer-guide/iso-installer.md (hunk 22 above)
Comment (danmcd) [reply docs/developer-guide/iso-installer.md:30]: I was being paranoid about both images AND about having sufficient swap for 16GB of RAM (our swap algorithm does 1-for-1 there until sufficiently large RAM).

File: docs/developer-guide/iso-installer.md (hunk 3 above)
Comment (danmcd) [reply docs/developer-guide/iso-installer.md:40]: Oh hell... will fix.

File: docs/developer-guide/iso-installer.md (hunk 4 above)
Comment (danmcd) [reply docs/developer-guide/iso-installer.md:47]: Good idea, will fix.

File: docs/developer-guide/zpool.md (hunk 23 above)
Comment (danmcd) [reply docs/developer-guide/zpool.md:42]: Probably, but I've not tried that yet.  Should do so on my headnode VM.

File: docs/developer-guide/zpool.md (hunk 5 above)
Comment (danmcd) [reply docs/developer-guide/zpool.md:48]: Uggh, will fix.

File: docs/developer-guide/zpool.md (hunk 6 above)
Comment (danmcd) [reply docs/developer-guide/zpool.md:70]: Hmm, it's not, but  I might've copied/pasted this from somewhere else that makes this mistake, will fix here.

File: docs/developer-guide/zpool.md (hunk 7 above)
Comment (danmcd) [reply docs/developer-guide/zpool.md:75]: Will fix.

File: docs/developer-guide/zpool.md (hunk 24 above)
Comment (danmcd) [reply docs/developer-guide/zpool.md:76]: Lemme take a swing at better language here.

File: docs/developer-guide/zpool.md (hunk 8 above)
Comment (danmcd) [reply docs/developer-guide/zpool.md:88]: Sure why not?

File: docs/developer-guide/zpool.md (hunk 9 above)
Comment (danmcd) [reply docs/developer-guide/zpool.md:93]: Yeah, thank you.

File: docs/operator-guide/server-setup.md (hunk 10 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:46]: Will fix.

File: docs/operator-guide/server-setup.md (hunk 11 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:438]: Will fix.

File: docs/operator-guide/server-setup.md (hunk 12 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:444]: Sure!

File: docs/operator-guide/server-setup.md (hunk 13 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:466]: Good idea.

File: docs/operator-guide/server-setup.md (hunk 14 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:471]: The latter of those two, sure.

File: docs/operator-guide/server-setup.md (hunk 15 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:472]: Yep.

File: docs/operator-guide/server-setup.md (hunk 16 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:474]: Yep!

File: docs/operator-guide/server-setup.md (hunk 17 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:478]: Yep!

File: docs/operator-guide/server-setup.md (hunk 18 above)
Comment (danmcd) [reply docs/operator-guide/server-setup.md:483]: Okay.

**Review by danmcd — COMMENTED** [review]
Lots of changes to make after dinner tonight.  Thank you @jasonbking and @papertigers !

File: docs/developer-guide/iso-installer.md (hunk 22 above)
Comment (danmcd) [reply docs/developer-guide/iso-installer.md:30]: Oh and also it needs room for Platform Images.

File: docs/developer-guide/ipxe-installer.md (hunk 21 above)
Comment (papertigers) [reply docs/developer-guide/ipxe-installer.md:38]: This can always be changed later

**Review by papertigers — APPROVED** [review]

File: docs/developer-guide/ipxe-installer.md (hunk 21 above)
Comment (danmcd) [reply docs/developer-guide/ipxe-installer.md:38]: Copied from other existing URLs.

**Review by jasonbking — APPROVED** [review]
